### PR TITLE
specify type of field when filling with todo!()

### PR DESCRIPTION
**Pull request by rosefromthedead**
i.e. after "Fill struct fields", instead of
```rust
S { a: todo!() }
```
now we can do
```rust
S { a: todo!("provide HashMap<String, Option<usize>>") }
```

### CODE REVIEW

**Comment by bjorn3** [comment]
This makes it harder to remove the `todo!()` again.

**Comment by rosehuds** [comment]
Yeah, that's true for editors without things like `c$` or helix's treesitter nav, but surely shift-end or a couple of ctrl-deletes isn't that bad? I could see it being a problem, but it's a matter of opinion whether it's worth it for the type hints.
Alternatively, does/could r-a provide correct inlay type hints for struct fields with broken initialisers? I think that would solve this problem without affecting the `todo!()`s at all.

**Comment by bjorn3** [comment]
> Alternatively, does/could r-a provide correct inlay type hints for struct fields with broken initialisers?

That could work I think. Alternatively there could be a setting to switch between the current and new `todo!()` placeholder.

**Comment by kpreid** [comment]
> Alternatively there could be a setting to switch between the current and new `todo!()` placeholder.

There is already the setting `rust-analyzer.assist.expressionFillDefault`, which could gain a new value like `todoWithType`.

**Comment by rosehuds** [comment]
IMO there should be a new setting - someone wanting type hints might also want defaults where possible

File: crates/rust-analyzer/src/config.rs [hunk 1]
```diff
@@ -56,6 +56,8 @@ config_data! {
     struct ConfigData {
         /// Placeholder expression to use for missing expressions in assists.
         assist_expressionFillDefault: ExprFillDefaultDef              = "\"todo\"",
+        /// Specify type of missing expression inside generated todo macro invocation.
+        assist_expressionFillTypeHints_enable: bool = "true",
```
Comment (Veykril) [inline crates/rust-analyzer/src/config.rs:60]: This should be disabled by default, this is a rather noisy setting so I don't think it makes a good default if enabled.

Comment (rosehuds) [reply crates/rust-analyzer/src/config.rs]: Done :)

**Comment by Veykril** [comment]
I'm sorry for not having merged this PR by now, but I was still contemplating whether having this is worth it or not, and I am inclined to say no after all.

> Alternatively, does/could r-a provide correct inlay type hints for struct fields with broken initialisers? I think that would solve this problem without affecting the todo!()s at all.

But I like this, I think we could maybe do the following which would help out with the main issue you want to solve which is to annotate `todo/unimplemented` macro calls with an  inlay hint showing the type that is expected?

**Comment by Veykril** [comment]
@kpreid what are your thoughts on my previous comment? (Since you were the one who opened the issue)

**Comment by kpreid** [comment]
> what are your thoughts on my previous comment regarding the use of inlay hints instead? 

Hm. In principle it is elegant, but I don't think it would work very well for the situations I imagine.

1. As far as I know, you can't copy text out of an inlay hint, so I would have to retype any relevant type names.
2. I have inlay hints set to `offUnlessPressed`, which means I can't look at them when I'm typing but would have to alternate looking and typing. (This is because they're a lot of visual noise, especially on any of the stock light themes, and they stretch lines past their usual length, creating awkward formatting.)

I haven't recently had very many complex struct literals to deal with, though, so my hypotheticals may not be very refined.

**Comment by flodiebold** [comment]
I wonder if we could show signature help for struct literals. I think that plus better expected-type-based completions would help here.

**Comment by Veykril** [comment]
> I wonder if we could show signature help for struct literals. I think that plus better expected-type-based completions would help here.

That sounds like a nice idea imo

**Comment by HKalbasi** [comment]
Another idea is showing types of `todo!()`s as an inlay hint inside their parenthesis.

**Pull request closed without merging**